filament/backend/src/metal/MetalDriver.mm

```diff
@@ -89,13 +89,19 @@
 }
 #endif
 
-void MetalDriver::beginFrame(int64_t monotonic_clock_ns, uint32_t frameId) {
+void MetalDriver::beginFrame(int64_t monotonic_clock_ns, uint32_t frameId,
+        backend::FrameFinishedCallback callback, void* user) {
     mContext->framePool = [[NSAutoreleasePool alloc] init];
 
     id<MTLCommandBuffer> commandBuffer = acquireCommandBuffer(mContext);
     [commandBuffer addCompletedHandler:^(id <MTLCommandBuffer> buffer) {
         mContext->resourceTracker.clearResources(buffer);
     }];
+
+    // If a callback was specified, then the client is responsible for presenting the frame.
+    mContext->clientPresent = callback != nullptr;
+    mContext->frameFinishedCallback = callback;
```

Comment:
do you need clientPresent? can't you use frameFinishedCallback? (whether it's null or not)